cub/test/catch2_test_device_select_flagged_if.cu

```diff
@@ -0,0 +1,286 @@
+/******************************************************************************
+ * Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#include <cub/device/device_select.cuh>
+
+#include <thrust/count.h>
+#include <thrust/logical.h>
+
+#include "catch2_test_helper.h"
+#include "catch2_test_launch_helper.h"
+
+template <class T, class FlagT, class Pred>
+static c2h::host_vector<T>
+get_reference(c2h::device_vector<T> const& in, c2h::device_vector<FlagT> const& flags, Pred if_predicate)
+{
+  struct selector
+  {
+    T const* ref_begin      = nullptr;
+    FlagT const* flag_begin = nullptr;
+    Pred const& if_pred;
+
+    constexpr selector(T const* ref, FlagT const* flag, Pred const& pred) noexcept
+        : ref_begin(ref)
+        , flag_begin(flag)
+        , if_pred(pred)
+    {}
+
+    bool operator()(T const& val) const
+    {
+      const auto pos = &val - ref_begin;
+      return static_cast<bool>(if_pred(flag_begin[pos]));
+    }
+  };
+
+  c2h::host_vector<T> reference   = in;
+  c2h::host_vector<FlagT> h_flags = flags;
+
+  const selector pred{
+    thrust::raw_pointer_cast(reference.data()), thrust::raw_pointer_cast(h_flags.data()), if_predicate};
+  const auto boundary = std::stable_partition(reference.begin(), reference.end(), pred);
+  reference.erase(boundary, reference.end());
+  return reference;
+}
+
```

Comment:
suggestion: Personally, I dislike the approach of taking `val` by reference and computing the offset by taking its address. How about computing the reference, using, for instance, a zip iterator? 
```suggestion

template<typename PredOpT>
struct predicate_op_wrapper_t
{
    PredOpT if_pred;
    template<typename FlagT, typename ItemT>
    __host__ __device__ bool operator()(thrust::tuple<FlagT, ItemT> tuple) const
    {
        const auto flag = thrust::get<0>(tuple);
        return static_cast<bool>(if_pred(flag));
    }
};

template <class T, class FlagT, class Pred>
static c2h::host_vector<T>
get_reference(c2h::device_vector<T> const& in, c2h::device_vector<FlagT> const& flags, Pred if_predicate)
{
    c2h::host_vector<T> reference   = in;
    c2h::host_vector<FlagT> h_flags = flags;
    // Zips flags and items
    auto zipped_in_it = thrust::make_zip_iterator(h_flags.cbegin(), reference.cbegin());

    // Discards the flags part and only keeps the items
    auto zipped_out_it = thrust::make_zip_iterator(thrust::make_discard_iterator(), reference.begin());

    auto end = std::copy_if(zipped_in_it, zipped_in_it + in.size(), zipped_out_it, predicate_op_wrapper_t<Pred>{if_predicate});
    reference.resize(thrust::distance(zipped_out_it, end));
    return reference;
}

```